### ISSUE TITLE
conditional events like in SBML

### ISSUE THREAD

**Issue by metelkin** [metrumresearchgroup/mrgsolve#881]
Hello,
I wonder if there is any way to use a conditional event in mrgsolve.
I mean here case when we want to update variable's or parameter's value depending on condition.

For example, I want to change `S1` variable when `S2 <= 5` at an unknown time point.
I cannot find it in docs. Sorry if it's been explained somewhere.

**Comment by kylebaron** [comment]
Hi - This is called `mtime` similar to nonmem:

https://mrgsolve.github.io/user_guide/model-specification.html#self.mtime

In the example, the time of the change is coded as a parameter (controllable from R) but you don't _have_ to know the change time before the simulation starts (it could be random or depend on how the model progresses up to some decision point).

**Comment by metelkin** [comment]
Sorry, I still don't understand this approach.
It seems like the `mtime` mechanism can be applied when you want to change event time as a parameter from R or calculate it by an expression. It is still unclear how the event can be run by some condition (which does not include time).

The example describes the updates of `KA` value. But what if I'd like to change the value of an ODE variable, to set it as zero, or something else.

I've prepared an example to explain what I mean.

```cpp
$PARAM
k1 = 0.01

$CMT
x1
x2

$PREAMBLE
double comp1 = 1.0;

$MAIN
x1_0 = 10.0;
x2_0 = 0.0;

// need for event
// when x1 <= 6, set x1 = 10

$ODE
double r1 = k1 * x1 * comp1;
dxdt_x1 = -r1;
dxdt_x2 = r1;
```

**Comment by kylebaron** [comment]
Sure ... you can do something like this:

It's  a little more work because you create an object with event type members.  Use evid 8 to replace the amount in a compartment. Each event needs a time and gets scheduled some time in the future. If you want something to happen `now` set that (it think the logic might need to be tweaked for cases when you set the time of the event to the current time ... need to check on that). Then push that object back so mrgsolve can implement it. 

There is documentation here: https://mrgsolve.github.io/user_guide/mtime.html#modeled-doses
I do use this stuff but not every day; so any suggestions on the docs and usage would be welcomed.


In this example
- initialize `depot` to 120
- reset `cent` to 100 any time `cent` drops below 40 


``` r
library(tidyverse)
library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter


code <- '
[ param ] KA = 1.5, V = 30, CL = 3

[ pkmodel ] cmt = "depot, cent", depot = TRUE

[ main ] 

depot_0 = 120;

if(cent < 40 && TIME > 3) {
  mrg::evdata ev(TIME, 8); 
  ev.amt = 100;
  ev.cmt = 2;
  ev.now = true;
  self.mevector.push_back(ev);
}
'

mod <- mcode("reset", code, end = 72, delta = 0.1)
#> Building reset ...
#> done.

mod %>% mrgsim() %>% plot()
```

![](https://i.imgur.com/3i24p9G.png)

<sup>Created on 2021-11-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by kylebaron** [comment]
I think this is your example:


``` r
library(tidyverse)
library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter
code <- '
$PARAM
k1 = 0.01

$CMT
x1
x2

$PREAMBLE
double comp1 = 1.0;

$MAIN
x1_0 = 10.0;
x2_0 = 0.0;

// need for event
// when x1 <= 6, set x1 = 10
if(x1 <= 6) {
  mrg::evdata ev(TIME, 8); 
  ev.amt = 10;
  ev.cmt = 1;
  ev.now = true;
  self.mevector.push_back(ev);
}


$ODE
double r1 = k1 * x1 * comp1;
dxdt_x1 = -r1;
dxdt_x2 = r1;
'

mod <- mcode("x1x2", code)
#> Building x1x2 ...
#> done.

mrgsim(mod, end = 480) %>% plot()
```

![](https://i.imgur.com/Hz0x5Bg.png)

<sup>Created on 2021-11-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by metelkin** [comment]
Thank you, @kylebaron.
I guess this is what I need. I should read the documentation more attentively.
I'll test the approach for more examples.

I guess I should also select the appropriate "hmax" parameter to have enough tolerance for "x1 <= 6" condition and to not skip some events.

**Comment by kylebaron** [comment]
That's a good point @metelkin .  

The code I posted will be dependent on input data design (or how frequently you ask for **output** records).  There are ways to get it to stop / look more frequently, but sort of hacky.  I **think** reducing `hmax` will help prevent overshoot, but I the observation design factors in too (and might be more important).

Anyway, there is no way to monitor `x1` values in the `[ode]` block and do this compartment reset and I think that makes sense. So if the system has to advance a long time (without stopping to log some of the data) this change could be imprecisely implemented.

**Comment by metelkin** [comment]
Hello, @kylebaron
I tested the approach for more cases. There are my comments:

1) When we use a more complex example, namely when the right side of statement `ev.amt=` depends on a dynamic variable I've got the wrong results. The maximal `x1` value here is about 25 but it must be about 16.

```r
code <- '
$PARAM @annotated
k1 : 0.01 : 

$CMT @annotated
x1_amt_ : as amount
x2_amt_ : as amount

$GLOBAL
#define x1 (x1_amt_ / comp1)
#define x2 (x2_amt_ / comp1)

$PREAMBLE
double comp1 = 1.0;

$MAIN
x2_amt__0 = (0.0) * comp1;
x1_amt__0 = (10.0) * comp1;

if (cond1>=0) {
  mrg::evdata ev_x1_(TIME, 8); 
  ev_x1_.amt = (x1 + 10.0) * comp1;
  ev_x1_.cmt = 1;
  ev_x1_.now = true;
  self.mevector.push_back(ev_x1_);
}

$ODE
double r1 = k1 * x1 * comp1;
double cond1 = 6.0 - x1;

dxdt_x1_amt_ = (-1)*r1;
dxdt_x2_amt_ = (1)*r1;

$CAPTURE
x1
x2
'
mod <- mcode("x1x2", code)

sim <- nameless_model %>%
  mrgsim(
    delta = 1,
    hmax = 0,
    maxsteps = 1e9,
    atol = 1e-7,
    rtol = 1e-4,
    end = 120
  )

# plot results
sim %>% plot(type='l')
```
![Rplot](https://user-images.githubusercontent.com/23167846/142604482-4760e96e-97e6-46dd-a3ad-b7d50d18f04c.png)

I've found a solution putting `if () {...}` code into `$TABLE` not `$MAIN` part of cpp file. This works fine.

**Comment by metelkin** [comment]
2) As you've mentioned the results also depend on output time points or `delta` parameter. `delta` should be lower than typical switching time. 
Hypothetically to estimate switching time more precisely we should monitor in `$ODE` block. As I see it's impossible because `self.mevector` is unavailable there.
Another option would be to have direct access to dynamic variables (compartments) inside `$ODE` block. Is it possible to mutate them without creating an event? 
Or maybe it is reasonable to be able to create a `mevent` at an integration point in `$ODE`.

**Comment by kylebaron** [comment]
>I've found a solution putting if () {...} code into $TABLE not $MAIN part of cpp file. This works fine.

That's right. I have to re-learn this lesson all the time: Monitor the state in `$TABLE` and make changes before going to the next record. I need to write that down somewhere.

I'm trying to think of ways to make this happen inside `$ODE`. Maybe you could do it with another solver. The problem is that we've already told the solver to advance from `T1 ` to `T2`.  Maybe you could set something up where we detect some condition and signal back to the ODE solver to stop, reinitialize at the new initials and keep advancing. You can't do that now and I'm not sure it would work as expected if we tried it (that rhs function gets called in a bunch of different circumstances).

**Comment by metelkin** [comment]
@kylebaron 
I guess I've got the preliminary solution for my problem. Thank you for your assistance.
In general, the only problem now is the preciseness of time points to calculate the user's conditions.
Hopefully, it will be enhanced in a future solver version.

As a result of the discussion, I implemented the support of events in mrgsolve export in heta-simulator. I've published the demo in  R markdown format here:  <http://shiny.insysbio.com/sbml-mrgsolve-case/>. This case converts online an SBML file with events into mrgsolve and runs default simulations. Looks like everything is working.

BTW in Matlab's and Julia's solvers, the specific mechanism for handling continuous events is used. It is a sort of step back in time to reach the required condition tolerance. I don't think it is very useful for mrgsolve users, just for information.
